This patch fixes communty bug 3877 
"4.2.8p16 client rejects server response with zero ppoll." and may be removed
when upgrading to a version of NTP with that bug fixed.

diff -Nru a/ntpd/ntp_proto.c b/ntpd/ntp_proto.c
--- ntpd/ntp_proto.c
+++ ntpd/ntp_proto.c
@@ -2306,37 +2306,29 @@
 	 * this maximum and advance the headway to give the sender some
 	 * headroom. Very intricate.
 	 */
+	peer->ppoll = max(peer->minpoll, pkt->ppoll);
 
 	/*
 	 * Check for any kiss codes. Note this is only used when a server
-	 * responds to a packet request.
+	 * responds to a client request.
 	 */
-
-	/*
-	 * Check to see if this is a RATE Kiss Code
-	 * Currently this kiss code will accept whatever valid poll
-	 * rate that the server sends
-	 */
-	if (   (NTP_MINPOLL > pkt->ppoll)
-	    || (NTP_MAXPOLL < pkt->ppoll)
-	   ) {
-		DPRINTF(2, ("RATEKISS: Invalid ppoll (%d) from %s\n",
-				pkt->ppoll, stoa(&rbufp->recv_srcadr)));
-		sys_badlength++;
-		return;			/* invalid packet poll */
-	}
-	peer->ppoll = max(peer->minpoll, pkt->ppoll);
 	if (kissCode == RATEKISS) {
+		if (   pkt->ppoll < NTP_MINPOLL
+		    || pkt->ppoll > NTP_MAXPOLL) {
+			DPRINTF(2, ("Ignoring ppoll %d RATE KoD from %s\n",
+				    pkt->ppoll, stoa(&rbufp->recv_srcadr)));
+			sys_badlength++;
+			return;			/* invalid packet poll */
+		}
 		peer->selbroken++;	/* Increment the KoD count */
 		report_event(PEVNT_RATE, peer, NULL);
-		if (pkt->ppoll > peer->minpoll)
-			peer->minpoll = peer->ppoll;
+		peer->minpoll = peer->ppoll;
 		peer->burst = peer->retry = 0;
 		peer->throttle = (NTP_SHIFT + 1) * (1 << peer->minpoll);
 		poll_update(peer, pkt->ppoll, 0);
 		return;				/* kiss-o'-death */
-	}
-	if (kissCode != NOKISS) {
+
+	} else if (kissCode != NOKISS) {
 		peer->selbroken++;	/* Increment the KoD count */
 		return;		/* Drop any other kiss code packets */
 	}
